modules/linear_boltzmann_solvers/lbs_solver/volumetric_source/volumetric_source.cc

```diff
@@ -0,0 +1,125 @@
+                  object_stack, params.GetParamValue<size_t>("logical_volume_handle"))
+              : nullptr),
+    strength_(params.GetParamVectorValue<double>("group_strength")),
+    function_(params.ParametersAtAssignment().Has("function_handle")
+                ? GetStackItemPtrAsType<SpatialMaterialFunction>(
+                    object_stack, params.GetParamValue<size_t>("function_handle"))
+                : nullptr)
+{
+  if (not logvol_ and block_ids_.empty())
+    throw std::invalid_argument("A volumetric source must be defined with a logical volume, "
+                                "block IDs, or both. Neither were specified.");
+  if (function_ and not strength_.empty())
+    throw std::invalid_argument(
+      "If a function is provided, the source strength should not be set.");
+}
+
+void
+lbs::VolumetricSource::Initialize(const LBSSolver& lbs_solver)
+{
+  // Set the source strength vector
+  if (not function_ and not strength_.empty())
+    if (strength_.size() != lbs_solver.NumGroups())
+      throw std::invalid_argument("The number of groups in the source strength vector must "
+                                  "match the number of groups in the solver the source is "
+                                  "attached to.");
+
+  // Set cell subscribers based on logical volumes, block IDs, or both
+  subscribers_.clear();
+  if (logvol_ and block_ids_.empty())
+  {
+    std::set<int> blk_ids;
+    for (const auto& cell : lbs_solver.Grid().local_cells)
+      if (logvol_->Inside(cell.centroid_))
+      {
+        blk_ids.insert(cell.material_id_);
+        subscribers_.push_back(cell.local_id_);
+      }
+  }
+  else if (not logvol_ and not block_ids_.empty())
+  {
+    for (const auto& cell : lbs_solver.Grid().local_cells)
+      if (std::find(block_ids_.begin(), block_ids_.end(), cell.material_id_) != block_ids_.end())
+        subscribers_.push_back(cell.local_id_);
+  }
+  else
+  {
+    for (const auto& cell : lbs_solver.Grid().local_cells)
+      if (logvol_->Inside(cell.centroid_) and
+          std::find(block_ids_.begin(), block_ids_.end(), cell.material_id_) != block_ids_.end())
+        subscribers_.push_back(cell.local_id_);
+  }
+
+  num_local_subsribers_ = subscribers_.size();
+  mpi_comm.all_reduce(num_local_subsribers_, num_global_subscribers_, mpi::op::sum<size_t>());
+
+  log.LogAll() << "Volumetric source has " << num_local_subsribers_
+               << " subscribing cells on processor " << opensn::mpi_comm.rank() << ".";
+  log.Log() << "Volumetric source has " << num_global_subscribers_ << " total subscribing cells.";
+}
+
+std::vector<double>
+lbs::VolumetricSource::operator()(const Cell& cell, const Vector3& xyz, const int num_groups) const
+{
+  if (std::count(subscribers_.begin(), subscribers_.end(), cell.local_id_) == 0)
+    return std::vector<double>(num_groups, 0.0);
+  else if (not function_)
+    return strength_.empty() ? std::vector<double>(num_groups, 1.0) : strength_;
```

Comment:
I do not think we should allow default options. If no functions are given and strength is empty, then complaint. I think the user must enter a value/array.

Reply:
This was the default behavior before this PR, which was used in some of the tests. I will make the change.